racket/collects/json/main.rkt

```diff
@@ -356,21 +367,29 @@
               (arithmetic-shift (hex-convert c3) 4)
               (hex-convert c4)))
          (define e (get-hex))
-         (define e*
-           (cond
-             [(<= #xD800 e #xDFFF)
-              (define (err-missing)
-                (err "bad string \\u escape, missing second half of a UTF-16 pair"))
-              (unless (eqv? (read-byte i) (char->integer #\\)) (err-missing))
-              (unless (eqv? (read-byte i) (char->integer #\u)) (err-missing))
-              (define e2 (get-hex))
-              (cond
-                [(<= #xDC00 e2 #xDFFF)
-                 (+ (arithmetic-shift (- e #xD800) 10) (- e2 #xDC00) #x10000)]
-                [else
-                 (err "bad string \\u escape, bad second half of a UTF-16 pair")])]
-             [else e]))
-         (keep-char (integer->char e*) result pos converter)]
+         (define-values (e* new-result new-pos)
+           (let resync ([e e] [result result] [pos pos])
+             (cond
+               [(<= #xD800 e #xDFFF)
+                (cond
+                  [(equal? (peek-bytes 2 0 i) #"\\u")
```

Comment:
Is this right when only the second half of a surrogate pair is present? It seems like it might try to read (another) second half and even generate a bad character if there are two second halves in a row, but maybe I'm missing something.

Reply:
You're right.  I've added more tests and adjusted the implementation to handle this.